### PR TITLE
Fix rounding of surplus/protocol_fee/score

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Related to https://github.com/cowprotocol/services/issues/2797#issuecomment-2264843339

Fixes issue in `driver` and new `autopilot` code. Old `autopilot` code is not fixed and I don't intend to do it. This means that settlement observations will still be wrong until new code is used. Dependent on https://github.com/cowprotocol/services/pull/2843

Note that this change affects the whole range of functionalities: surplus, protocol fee, score calculation.

# Changes

- [ ] Whenever we calculate `buy_amount` for `sell` orders, use `ceil_div` to be consistent with settlement contract logic 

## How to test
Added unit test that shows that new code calculates the expected circuit breaker surplus of `384509480572312`
Expected score is 2 x surplus.

### CODE REVIEW

**Comment by fhenneke** [comment]
The circuit breaker reported a score of `769018961144626` (the driver would probably report 1 or 2 wei less due to rounding down final results.)

If the driver now reports that score, it should be fine for now.

With regards to the difference of old and new autopilot: we do not really depend on surplus from the settlement_observations table, so it should be fine if that number is off for now. (We do sync it to dune but not being super accurate there should be fine. It there any other significance of the surplus/protocol fee/score computation in the autopilot?

**Comment by sunce86** [comment]
> If the driver now reports that score, it should be fine for now.

Autopilot score checks out with your example 🥳  I would expect `driver` is the same as the logic is same (unless we have a bug somewhere).

**Comment by sunce86** [comment]
@fhenneke

Why circuit breaker returns `769018961144626`? 

surplus is 121 and protocol fee is also 121 so it's 2 * surplus = `2* 384509480572312 = 769018961144624` and that's how much autopilot now calculates?

File: crates/autopilot/src/domain/settlement/solution/mod.rs
Comment (sunce86) [inline crates/autopilot/src/domain/settlement/solution/mod.rs:319]: will remove test before merging

**Comment by sunce86** [comment]
> Rounding before or after summation can lead to differences in the final result.

This doesn't sound right. If surplus is X, it should contribute to score (which is surplus + protocol fees) with the same amount X and not X+1.

Now, whether we round down or round up is a matter of choice. Backend does the rounding down by standard everywhere, UNLESS there is a specific need to round up, like we have in this PR to be consistent with settlement contract implementation.

I would suggest to go with this PR as default behaviour. How problematic is for circuit breaker to align with this?

File: crates/autopilot/src/domain/settlement/solution/mod.rs
Comment (fleupold) [reply crates/autopilot/src/domain/settlement/solution/mod.rs:319]: I think you can even leave this checked in as a regression test if it's properly named.

**Review by fleupold — COMMENTED** [review]
> This doesn't sound right.

Maybe not logically, but at least mathematically `ceil(5.6 + 5.6) = 11 != ceil(5.6) + ceil(5.6) == 12`)

The pure code change looks good to me but will defer to @fhenneke subject expertise to weigh in if this rounding solves the problem.

**Comment by fhenneke** [comment]
Rounding limit amounts should be correct with this PR.

Rounding surplus and protocol fees individually or after adding them are both fine. The circuit breaker can do both. The differences in scores should be double digits in wei. Using one-sided testing for blacklisting will not cause an issue even if we do not change the circuit breaker.

There are additional issues with rounding in the driver/autopilot. With the proposed changes, the driver should never overestimate scores. I do not think, however, that the implementation of the driver should be regarded as ground truth regarding scores.

**Comment by sunce86** [comment]
> There are additional issues with rounding in the driver/autopilot.

Let's merge this PR as it also addresses some inconsistencies/bugs in the code (sometime using ceil sometime not). Will have to go back to original issue to try to detect these additional rounding issues you refer to.

File: crates/autopilot/src/domain/settlement/solution/mod.rs
Comment (sunce86) [reply crates/autopilot/src/domain/settlement/solution/mod.rs]: fixed

**Review by fleupold — APPROVED** [review]

**Review by Mateo-mro — APPROVED** [review]

**Pull request merged**
